src/main/java/ch/njol/skript/entity/MinecartData.java

```diff
@@ -12,142 +11,116 @@
 import org.bukkit.entity.minecart.StorageMinecart;
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
-
 import ch.njol.skript.lang.Literal;
 import ch.njol.skript.lang.SkriptParser.ParseResult;
-import ch.njol.skript.util.Utils;
 import ch.njol.skript.variables.Variables;
 
-/**
- * @author Peter Güttinger
- */
 public class MinecartData extends EntityData<Minecart> {
-	
-	@SuppressWarnings("null")
+
 	private static enum MinecartType {
-		ANY(Minecart.class, "minecart"),
 		NORMAL(RideableMinecart.class, "regular minecart"),
 		STORAGE(StorageMinecart.class, "storage minecart"),
 		POWERED(PoweredMinecart.class, "powered minecart"),
 		HOPPER(HopperMinecart.class, "hopper minecart"),
 		EXPLOSIVE(ExplosiveMinecart.class, "explosive minecart"),
 		SPAWNER(SpawnerMinecart.class, "spawner minecart"),
 		COMMAND(CommandMinecart.class, "command minecart");
-		
-		@Nullable
-		final Class<? extends Minecart> c;
+
+		private final Class<? extends Minecart> entityClass;
 		private final String codeName;
-		
-		MinecartType(final @Nullable Class<? extends Minecart> c, final String codeName) {
-			this.c = c;
+
+		MinecartType(Class<? extends Minecart> entityClass, String codeName) {
+			this.entityClass = entityClass;
 			this.codeName = codeName;
 		}
-		
+
 		@Override
 		public String toString() {
 			return codeName;
 		}
-		
+
 		public static String[] codeNames;
 		static {
-			final ArrayList<String> cn = new ArrayList<>();
-			for (final MinecartType t : values()) {
-				if (t.c != null)
-					cn.add(t.codeName);
+			var names = new ArrayList<>();
+			names.add("minecart");
+			for (MinecartType type : values()) {
+				names.add(type.codeName);
 			}
-			codeNames = cn.toArray(new String[0]);
+			codeNames = names.toArray(new String[0]);
 		}
 	}
-	
+
 	static {
 		EntityData.register(MinecartData.class, "minecart", Minecart.class, 0, MinecartType.codeNames);
-		
 		Variables.yggdrasil.registerSingleClass(MinecartType.class, "MinecartType");
 	}
-	
-	private MinecartType type = MinecartType.ANY;
-	
+
+	private MinecartType type = MinecartType.NORMAL;
+	private boolean isSupertype = true;
+
 	public MinecartData() {}
-	
-	public MinecartData(final MinecartType type) {
+
+	public MinecartData(MinecartType type) {
```

Comment:
Is this called for deserialization? E.g. if I do `set {minecart} to a minecart` it would actually deserialize as `a regular minecart`

Reply:
Not when a type is provided in the constructor, that's the point of the constructor.